imap/jmap_mail.c

```diff
@@ -11644,26 +11644,64 @@ static int _email_bulkupdate_plan(struct email_bulkupdate *bulk, ptrarray_t *upd
             int r = mailbox_quota_check(plan->mbox, qdiffs);
             if (r) {
                 _email_updateplan_error(plan, r, bulk->set_errors);
-                strarray_append(&erroneous_plans, plan->mbox_id);
+                hash_insert(plan->mbox_id, (void*)IMAP_QUOTA_EXCEEDED, &erroneous_plans);
             }
         }
     }
-    hash_iter_free(&iter);
+    if (hash_numrecords(&erroneous_plans)) {
+        /* Fail any update where the email is an erroneous mailbox plan */
+        hash_iter_reset(plan_iter);
+        while (hash_iter_next(plan_iter)) {
+            struct email_updateplan *plan = hash_iter_val(plan_iter);
+            if (hash_lookup(plan->mbox_id, &erroneous_plans))
+                continue;
 
-    /* Remove erroneous plans */
-    for (i = 0; i < strarray_size(&erroneous_plans); i++) {
-        const char *mbox_id = strarray_nth(&erroneous_plans, i);
-        struct email_updateplan *plan = hash_del(mbox_id, &bulk->plans_by_mbox_id);
-        if (!plan) continue;
-        jmap_closembox(bulk->req, &plan->mbox);
-        _email_updateplan_free_p(plan);
+            int i;
+            for (i = 0; i < ptrarray_size(&plan->copy); i++) {
+                ptrarray_t *uidrecs = ptrarray_nth(&plan->copy, i);
+                int j;
+                for (j = 0; j < ptrarray_size(uidrecs); j++) {
+                    struct email_uidrec *uidrec = ptrarray_nth(uidrecs, j);
+                    intptr_t errcode = (intptr_t)hash_lookup(uidrec->mboxrec->mbox_id, &erroneous_plans);
+                    if (errcode && !json_object_get(bulk->set_errors, uidrec->email_id)) {
+                        json_object_set_new(bulk->set_errors, uidrec->email_id,
+                                jmap_server_error(errcode));
+                    }
+                }
+            }
+            ptrarray_t *uidrecs[3] = { &plan->setflags, &plan->delete, &plan->snooze };
```

Comment:
Why is plan->copy handled differently above and not part of this loop? It looks the same to me

Reply:
Not quite. plan->copy is a list of a list of uidrecs. The others are list of records.